File: python/test/lib/util_test.py
# Copyright 2015 ETH Zurich
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
:mod:`lib_util_test` --- lib.util unit tests
=====================================================
"""
# Stdlib
import builtins
from unittest.mock import patch, mock_open

# External packages
import nose
import nose.tools as ntools
import yaml

# SCION
from python.lib.errors import (
    SCIONIOError,
    SCIONYAMLError,
)
from python.lib.util import (
    load_yaml_file,
    write_file,
)


class TestWriteFile(object):
    """
    Unit tests for lib.util.write_file
    """
    @patch("lib.util.os.rename", autospec=True)
    @patch.object(builtins, 'open', mock_open())
    @patch("lib.util.os.makedirs", autospec=True)
    @patch("lib.util.os.path.dirname", autospec=True)
    def test_basic(self, dirname, makedirs, rename):
        dirname.return_value = "Dir_Name"
        # Call
        write_file("File_Path", "Text")
        # Tests
        dirname.assert_called_once_with("File_Path")
        makedirs.assert_called_once_with("Dir_Name", exist_ok=True)
        builtins.open.assert_called_once_with("File_Path.new", 'w')
        builtins.open.return_value.write.assert_called_once_with("Text")
        rename.assert_called_once_with("File_Path.new", "File_Path")

    @patch("lib.util.os.makedirs", autospec=True)
    def test_mkdir_error(self, mkdir):
        mkdir.side_effect = FileNotFoundError
        # Call
        ntools.assert_raises(SCIONIOError, write_file, "File_Path", "Text")

    @patch.object(builtins, 'open', mock_open())
    @patch("lib.util.os.makedirs", autospec=True)
    def test_file_error(self, mkdir):
        builtins.open.side_effect = PermissionError
        # Call
        ntools.assert_raises(SCIONIOError, write_file, "File_Path", "Text")

    @patch("lib.util.os.rename", autospec=True)
    @patch.object(builtins, 'open', mock_open())
    @patch("lib.util.os.makedirs", autospec=True)
    def test_rename_error(self, mkdir, rename):
        rename.side_effect = PermissionError
        # Call
        ntools.assert_raises(SCIONIOError, write_file, "File_Path", "Text")


class Loader(object):
    """
    Helper class for load_yaml_file tests.
    """
    @patch.object(builtins, 'open', mock_open())
    def _file_error(self, target):
        builtins.open.side_effect = IsADirectoryError
        ntools.assert_raises(SCIONIOError, target, "File_Path")

    @patch.object(builtins, 'open', mock_open())
    def _check_loader_error(self, target, loader_path, excp, expected):
        with patch(loader_path, autospec=True) as loader:
            loader.side_effect = excp
            ntools.assert_raises(expected, target, "File_Path")


class TestLoadYAMLFile(Loader):
    """
    Unit tests for lib.util.load_yaml_file
    """
    def test_file_error(self):
        self._file_error(load_yaml_file)

    def test_json_error(self):
        for excp in (yaml.scanner.ScannerError, ):
            yield (
                self._check_loader_error, load_yaml_file, "lib.util.yaml.load",
                excp, SCIONYAMLError,
            )


if __name__ == "__main__":
    nose.run(defaultTest=__name__)

File: go/lib/pathmgr/util_test.go
// Copyright 2018 ETH Zurich
// Copyright 2019 ETH Zurich, Anapaya Systems
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package pathmgr_test

import (
	"strconv"
	"strings"
	"testing"

	"github.com/golang/mock/gomock"

	"github.com/scionproto/scion/go/lib/addr"
	"github.com/scionproto/scion/go/lib/common"
	"github.com/scionproto/scion/go/lib/snet"
	"github.com/scionproto/scion/go/lib/snet/mock_snet"
	"github.com/scionproto/scion/go/lib/xtest"
)

func buildSDAnswer(t testing.TB, ctrl *gomock.Controller, pathStrings ...string) []snet.Path {
	var paths []snet.Path
	for _, path := range pathStrings {
		paths = append(paths, createPath(t, ctrl, path))
	}
	return paths
}

func createPath(t testing.TB, ctrl *gomock.Controller, desc string) snet.Path {
	parts := strings.Split(desc, " ")
	path := mock_snet.NewMockPath(ctrl)
	interfaces := make([]snet.PathInterface, 0, len(parts))
	for _, part := range parts {
		tokens := strings.Split(part, "#")
		if len(tokens) != 2 {
			t.Fatalf("Invalid path description: %s", desc)
		}
		interfaces = append(interfaces, intf{
			ia: xtest.MustParseIA(tokens[0]),
			id: mustIfID(t, tokens[1]),
		})
	}
	path.EXPECT().Interfaces().Return(interfaces).AnyTimes()
	return path
}

func mustIfID(t testing.TB, s string) common.IFIDType {
	ifID, err := strconv.ParseUint(s, 10, 64)
	if err != nil {
		t.Fatalf("Failed to parse interface: %s", s)
	}
	return common.IFIDType(ifID)
}

type intf struct {
	ia addr.IA
	id common.IFIDType
}

func (i intf) IA() addr.IA         { return i.ia }
func (i intf) ID() common.IFIDType { return i.id }
